Stress test

dkhokhlov opened the pull request:
Stress test:
- produces configurable number of log files and lines (limited by ulimit)
- acts as ingester endpoint for agent (0.0.0.0:7080)
- checks for duplicate and lost lines
 - prints periodic reports until all lines received from agent
- supports all agent versions
- supports gzip compression
- does not override existing log files, appends by default
- todo: add https support

1. Running from shell

1.1 start agent with env vars:
```
export LOGDNA_LOG_DIRS=/home/dmitri/SOURCE/TMP/root/test
export LOGDNA_HOST=localhost:7080
export LOGDNA_USE_SSL=false
```

1.2 start test
```
$ python utils/stress_test/stress_test.py /home/dmitri/SOURCE/TMP/root/test 10 5000
2023-10-19 22:35:27 INFO   starting ingestor web server
 * Serving Flask app 'stress_test'
 * Debug mode: off
2023-10-19 22:35:33 INFO
2023-10-19 22:35:33 INFO   total seq:           10
2023-10-19 22:35:33 INFO   received seq:        10
2023-10-19 22:35:33 INFO   completed seq:       10
2023-10-19 22:35:33 INFO   total lines:         50000
2023-10-19 22:35:33 INFO   committed lines:     50000
2023-10-19 22:35:33 INFO   received lines:      50000
2023-10-19 22:35:33 INFO   duplicate lines:     0
2023-10-19 22:35:33 INFO   unrecognized lines:  0
2023-10-19 22:35:33 INFO   committed line rate: 8341 per sec
2023-10-19 22:35:33 INFO   received line rate:  8643 per sec
2023-10-19 22:35:33 INFO   received line bytes: 3939 KB
2023-10-19 22:35:33 INFO   run time:            6 sec
2023-10-19 22:35:33 INFO   FINISHED in 6 sec
```

2. Running in Docker

2.1 build test image
```
cd utils/stress-test
docker build -t  logdna-agent-stress-test
```

2.2 run test and agent images
```
docker network create stress_test
docker run --rm --net stress_test --name stress-test -v $(pwd)/logs:/var/log logdna-agent-stress-test:latest /var/log 50 1000000
docker run -it --net stress_test -e MZ_INGESTION_KEY=blah -e LOGDNA_LOG_DIRS=/var/log -e LOGDNA_HOST=stress-test:7080 -e LOGDNA_USE_SSL=false -it -v $(pwd)/logs:/var/log logdna/logdna-agent:3.9.0-dev
```

4. Help
```
$ python utils/stress_test/stress_test.py -h
usage: stress_test.py [-h] [--line_rate LINE_RATE] [--override] log_dir num_log_files num_lines

Agent Stress Test

positional arguments:
  log_dir               Directory where log files are stored.
  num_log_files         Number of log files to use.
  num_lines             Number of lines to add to each log file.

optional arguments:
  -h, --help            show this help message and exit
  --line_rate LINE_RATE
                        Line rate per log file. (default: 100)
  --override            Override existing log files. (default: False)
```

teriyakichild commented on utils/stress_test/stress_test.py:
```diff
@@ -0,0 +1,225 @@
+                num_duplicate_lines = (
+                    num_duplicate_lines + seq.total_lines - len(set(seq.line_ids))
+                )
+    line_rate = num_total_lines / run_time
```
line_rate should be `num_received_lines / run_time`
```suggestion
    line_rate = num_received_lines / run_time
```

dkhokhlov commented:
> I'm testing this in docker and it seems to be very slow:
> 
> ```
> docker network create stress_test
> docker run --rm --net stress_test --name stress-test -v $(pwd)/logs:/var/log logdna-agent-stress-test:latest /var/log 50 1000000
> docker run -it --net stress_test -e MZ_INGESTION_KEY=blah -e LOGDNA_LOG_DIRS=/var/log -e LOGDNA_HOST=stress-test:7080 -e LOGDNA_USE_SSL=false -it -v $(pwd)/logs:/var/log logdna/logdna-agent:3.9.0-dev
> ```
> 
> ```
> 2023-10-19 13:57:12 INFO   total seq:           50
> 2023-10-19 13:57:12 INFO   received seq:        50
> 2023-10-19 13:57:12 INFO   completed seq:       0
> 2023-10-19 13:57:12 INFO   total lines:         50000000
> 2023-10-19 13:57:12 INFO   received lines:      30074
> 2023-10-19 13:57:12 INFO   duplicate lines:     0
> 2023-10-19 13:57:12 INFO   unrecognized lines:  0
> 2023-10-19 13:57:12 INFO   line rate:           497 per sec
> 2023-10-19 13:57:12 INFO   run time:            60 sec
> ```

limited by line creation rate - by delay in the writer loop https://github.com/logdna/logdna-agent-v2/pull/572/files#diff-a3b59d908d4e8b6e9d752b3f48add84059bc02c16e1aaec1a99a39559959ca50R172

50 * 10 lines p/s  * 50 = 500

i was thinking number of files should scale it.  parameterize the delay as line rate per file?

teriyakichild commented:
ahh, I missed that.  Making that configurable sounds like a good idea because you may not want to keep scaling out the number of files.

dkhokhlov commented:
@teriyakichild 
ready for review
number of improvements:
- line_rate option, override_files option, README
- added committed line reports from writers, committed line rate
- added received line bytes, fixed committed lines in report

teriyakichild commented:
@dkhokhlov this looks good.  I do see these errors in the logdna-agent when running the stress-test with 30 or more log files: 

```
2023-10-20T13:32:26.607403Z  WARN notify_stream: unexpected event, kind: Other, from_path: None, to_path: None, attrs: EventAttributes { inner: Some(EventAttributesInner { tracker: None, flag: Some(Rescan), info: None, source: None, process_id: None }) }
2023-10-20T13:32:28.763857Z  WARN notify_stream: unexpected event, kind: Other, from_path: None, to_path: None, attrs: EventAttributes { inner: Some(EventAttributesInner { tracker: None, flag: Some(Rescan), info: None, source: None, process_id: None }) }
2023-10-20T13:32:30.817712Z  WARN notify_stream: unexpected event, kind: Other, from_path: None, to_path: None, attrs: EventAttributes { inner: Some(EventAttributesInner { tracker: None, flag: Some(Rescan), info: None, source: None, process_id: None }) }
2023-10-20T13:32:45.682710Z  WARN notify_stream: unexpected event, kind: Other, from_path: None, to_path: None, attrs: EventAttributes { inner: Some(EventAttributesInner { tracker: None, flag: Some(Rescan), info: None, source: None, process_id: None }) }
```

Is this expected?  Are we encroaching on the performance limits of the agent?

dkhokhlov commented:
> @dkhokhlov this looks good. I do see these errors in the logdna-agent when running the stress-test with 30 or more log files:
> 
> ```
> 2023-10-20T13:32:26.607403Z  WARN notify_stream: unexpected event, kind: Other, from_path: None, to_path: None, attrs: EventAttributes { inner: Some(EventAttributesInner { tracker: None, flag: Some(Rescan), info: None, source: None, process_id: None }) }
> 
> Is this expected? Are we encroaching on the performance limits of the agent?

This warning means that 3rd party component that provides file notifications to agent has experienced an overrun.  I saw this msg too under stress load close to the end of the test. Amazingly it did not affect the test. This is a subject for further testing. We did not see this msg before. Means stress test works. :)

teriyakichild commented:
okay cool, that makes sense.  I see some of the checks are failing so I will approve once those are fixed.
